File: Consumer Purchase Prediction using Machine Learning/scripts/python/eda.py
"""
Exploratory Data Analysis Script
Consumer Purchase Prediction Project
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Set visualization style
plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette("husl")

def load_data(file_path):
    """Load the dataset"""
    df = pd.read_csv(file_path)
    print(f"Dataset loaded successfully! Shape: {df.shape}")
    return df

def data_overview(df):
    """Perform initial data overview"""
    print("\n" + "="*50)
    print("DATA OVERVIEW")
    print("="*50)
    print(f"\nFirst 5 rows:")
    print(df.head())
    print(f"\nDataset Info:")
    print(df.info())
    print(f"\nStatistical Summary:")
    print(df.describe())
    print(f"\nMissing Values:")
    print(df.isnull().sum())
    print(f"\nDuplicate Rows: {df.duplicated().sum()}")

def analyze_target_variable(df):
    """Analyze target variable distribution"""
    print("\n" + "="*50)
    print("TARGET VARIABLE ANALYSIS")
    print("="*50)
    purchased_counts = df['Purchased'].value_counts()
    print(f"\nPurchased Distribution:")
    print(purchased_counts)
    print(f"\nPercentage:")
    print((purchased_counts / len(df)) * 100)
    
    # Visualization
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))
    axes[0].bar(purchased_counts.index, purchased_counts.values, color=['skyblue', 'coral'])
    axes[0].set_xlabel('Purchased')
    axes[0].set_ylabel('Count')
    axes[0].set_title('Purchased Distribution')
    axes[0].set_xticks([0, 1])
    axes[0].set_xticklabels(['No', 'Yes'])
    
    axes[1].pie(purchased_counts.values, labels=['No', 'Yes'], autopct='%1.1f%%', startangle=90)
    axes[1].set_title('Purchased Distribution (Pie Chart)')
    plt.tight_layout()
    plt.savefig('../../output/target_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()

def analyze_numerical_variables(df):
    """Analyze numerical variables"""
    print("\n" + "="*50)
    print("NUMERICAL VARIABLES ANALYSIS")
    print("="*50)
    numeric_cols = ['Age', 'EstimatedSalary']
    
    fig, axes = plt.subplots(2, 2, figsize=(15, 10))
    for idx, col in enumerate(numeric_cols):
        axes[idx, 0].hist(df[col], bins=30, edgecolor='black', alpha=0.7)
        axes[idx, 0].set_title(f'{col} Distribution (Histogram)')
        axes[idx, 0].set_xlabel(col)
        axes[idx, 0].set_ylabel('Frequency')
        
        axes[idx, 1].boxplot(df[col], vert=True)
        axes[idx, 1].set_title(f'{col} Distribution (Box Plot)')
        axes[idx, 1].set_ylabel(col)
    
    plt.tight_layout()
    plt.savefig('../../output/numerical_distributions.png', dpi=300, bbox_inches='tight')
    plt.close()

def analyze_categorical_variables(df):
    """Analyze categorical variables"""
    print("\n" + "="*50)
    print("CATEGORICAL VARIABLES ANALYSIS")
    print("="*50)
    gender_counts = df['Gender'].value_counts()
    print(f"\nGender Distribution:")
    print(gender_counts)
    print(f"\nPercentage:")
    print((gender_counts / len(df)) * 100)
    
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))
    axes[0].bar(gender_counts.index, gender_counts.values, color=['lightblue', 'lightpink'])
    axes[0].set_xlabel('Gender')
    axes[0].set_ylabel('Count')
    axes[0].set_title('Gender Distribution')
    
    axes[1].pie(gender_counts.values, labels=gender_counts.index, autopct='%1.1f%%', startangle=90)
    axes[1].set_title('Gender Distribution (Pie Chart)')
    plt.tight_layout()
    plt.savefig('../../output/categorical_distributions.png', dpi=300, bbox_inches='tight')
    plt.close()

def analyze_relationships(df):
    """Analyze relationships between variables"""
    print("\n" + "="*50)
    print("RELATIONSHIP ANALYSIS")
    print("="*50)
    
    # Age vs Purchased
    fig, axes = plt.subplots(1, 2, figsize=(15, 5))
    sns.boxplot(data=df, x='Purchased', y='Age', ax=axes[0])
    axes[0].set_title('Age Distribution by Purchase Status')
    axes[0].set_xticklabels(['No', 'Yes'])
    
    sns.boxplot(data=df, x='Purchased', y='EstimatedSalary', ax=axes[1])
    axes[1].set_title('Salary Distribution by Purchase Status')
    axes[1].set_xticklabels(['No', 'Yes'])
    plt.tight_layout()
    plt.savefig('../../output/relationships.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Scatter plot
    plt.figure(figsize=(12, 8))
    scatter = plt.scatter(df['Age'], df['EstimatedSalary'], 
                         c=df['Purchased'], cmap='coolwarm', 
                         alpha=0.6, s=50, edgecolors='black')
    plt.colorbar(scatter, label='Purchased')
    plt.xlabel('Age')
    plt.ylabel('Estimated Salary')
    plt.title('Age vs Estimated Salary (colored by Purchase Status)')
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig('../../output/scatter_plot.png', dpi=300, bbox_inches='tight')
    plt.close()

def correlation_analysis(df):
    """Perform correlation analysis"""
    print("\n" + "="*50)
    print("CORRELATION ANALYSIS")
    print("="*50)
    numeric_df = df[['Age', 'EstimatedSalary', 'Purchased']]
    correlation_matrix = numeric_df.corr()
    print(f"\nCorrelation Matrix:")
    print(correlation_matrix)
    print(f"\nCorrelation with Purchased:")
    print(correlation_matrix['Purchased'].sort_values(ascending=False))
    
    plt.figure(figsize=(10, 8))
    sns.heatmap(correlation_matrix, annot=True, fmt='.2f', cmap='coolwarm', 
                center=0, square=True, linewidths=1, cbar_kws={"shrink": 0.8})
    plt.title('Correlation Matrix')
    plt.tight_layout()
    plt.savefig('../../output/correlation_matrix.png', dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function to run EDA"""
    # Create output directory
    import os
    os.makedirs('../../output', exist_ok=True)
    
    # Load data
    df = load_data('../../data/Advertisement.csv')
    
    # Perform analysis
    data_overview(df)
    analyze_target_variable(df)
    analyze_numerical_variables(df)
    analyze_categorical_variables(df)
    analyze_relationships(df)
    correlation_analysis(df)
    
    print("\n" + "="*50)
    print("EDA COMPLETED SUCCESSFULLY!")
    print("="*50)

if __name__ == "__main__":
    main()

File: Consumer Purchase Prediction using Machine Learning/scripts/r/eda.R
# Exploratory Data Analysis Script
# Consumer Purchase Prediction Project

# Load necessary libraries
library(dplyr)
library(ggplot2)
library(corrplot)
library(VIM)

# Function to find project root by looking for data directory
find_project_root <- function() {
  # Start from current working directory
  current_dir <- getwd()
  max_levels <- 10
  project_marker <- file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "data", "Advertisement.csv")
  
  for (i in 1:max_levels) {
    # Check for the data file in nested structure
    if (file.exists(file.path(current_dir, project_marker))) {
      return(current_dir)
    }
    # Check for data file in current directory structure
    if (file.exists(file.path(current_dir, "data", "Advertisement.csv"))) {
      return(current_dir)
    }
    # Check if we're in the Consumer Purchase Prediction directory
    if (basename(current_dir) == "Consumer Purchase Prediction") {
      if (file.exists(file.path(current_dir, "Consumer Purchase Prediction", "data", "Advertisement.csv"))) {
        return(current_dir)
      }
      if (file.exists(file.path(current_dir, "data", "Advertisement.csv"))) {
        return(current_dir)
      }
    }
    # Move up one level
    parent_dir <- dirname(current_dir)
    if (parent_dir == current_dir) break  # Reached root
    current_dir <- parent_dir
  }
  return(NULL)
}

# Set working directory to project root
project_root <- find_project_root()
if (!is.null(project_root)) {
  setwd(project_root)
  cat("Working directory set to:", getwd(), "\n")
} else {
  cat("Warning: Could not find project root. Using current directory:", getwd(), "\n")
}

# Load the dataset - try multiple possible paths
data_paths <- c(
  file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "data", "Advertisement.csv"),
  file.path("data", "Advertisement.csv"),
  "Advertisement.csv"
)

data_path <- NULL
for (path in data_paths) {
  if (file.exists(path)) {
    data_path <- path
    break
  }
}

if (is.null(data_path)) {
  stop(paste("Cannot find Advertisement.csv. Searched in:\n",
             paste("  -", data_paths, collapse = "\n"),
             "\nCurrent working directory:", getwd()))
}

df <- read.csv(data_path, stringsAsFactors = TRUE)

cat("Dataset loaded successfully from:", data_path, "\n")
cat("Shape:", nrow(df), "rows,", ncol(df), "columns\n\n")

# 1. Data Overview
cat(paste(rep("=", 50), collapse=""), "\n")
cat("DATA OVERVIEW\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

cat("First 5 rows:\n")
print(head(df, 5))

cat("\nDataset Structure:\n")
str(df)

cat("\nStatistical Summary:\n")
print(summary(df))

cat("\nMissing Values:\n")
print(colSums(is.na(df)))

cat("\nDuplicate Rows:", sum(duplicated(df)), "\n\n")

# 2. Target Variable Analysis
cat(paste(rep("=", 50), collapse=""), "\n")
cat("TARGET VARIABLE ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

purchased_counts <- table(df$Purchased)
cat("Purchased Distribution:\n")
print(purchased_counts)
cat("\nPercentage:\n")
print(prop.table(purchased_counts) * 100)

# Create output directory if it doesn't exist
output_paths <- c(
  file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "output"),
  "output"
)

output_dir <- NULL
for (path in output_paths) {
  if (dir.exists(path)) {
    output_dir <- path
    break
  }
}

if (is.null(output_dir)) {
  # Create the first possible output directory
  output_dir <- output_paths[1]
  if (!dir.exists(dirname(output_dir))) {
    dir.create(dirname(output_dir), recursive = TRUE)
  }
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  cat("Created output directory:", output_dir, "\n")
}

# Visualize target variable
png(file.path(output_dir, "target_distribution_r.png"), width = 1200, height = 600, res = 300)
par(mfrow = c(1, 2))
barplot(purchased_counts, names.arg = c("No", "Yes"), 
        col = c("skyblue", "coral"),
        main = "Purchased Distribution (Bar Chart)",
        xlab = "Purchased", ylab = "Count")
pie(purchased_counts, labels = c("No", "Yes"), 
    main = "Purchased Distribution (Pie Chart)",
    col = c("skyblue", "coral"))
dev.off()

# 3. Numerical Variables Analysis
cat("\n", paste(rep("=", 50), collapse=""), "\n")
cat("NUMERICAL VARIABLES ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

png(file.path(output_dir, "numerical_distributions_r.png"), width = 1500, height = 1000, res = 300)
par(mfrow = c(2, 2))
hist(df$Age, main = "Age Distribution (Histogram)", xlab = "Age", col = "lightblue", breaks = 30)
boxplot(df$Age, main = "Age Distribution (Box Plot)", ylab = "Age", col = "lightblue")
hist(df$EstimatedSalary, main = "Estimated Salary Distribution (Histogram)", 
     xlab = "Estimated Salary", col = "lightgreen", breaks = 30)
boxplot(df$EstimatedSalary, main = "Estimated Salary Distribution (Box Plot)", 
        ylab = "Estimated Salary", col = "lightgreen")
dev.off()

# 4. Categorical Variables Analysis
cat(paste(rep("=", 50), collapse=""), "\n")
cat("CATEGORICAL VARIABLES ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

gender_counts <- table(df$Gender)
cat("Gender Distribution:\n")
print(gender_counts)
cat("\nPercentage:\n")
print(prop.table(gender_counts) * 100)

png(file.path(output_dir, "categorical_distributions_r.png"), width = 1200, height = 600, res = 300)
par(mfrow = c(1, 2))
barplot(gender_counts, col = c("lightblue", "lightpink"),
        main = "Gender Distribution", xlab = "Gender", ylab = "Count")
pie(gender_counts, labels = names(gender_counts),
    main = "Gender Distribution (Pie Chart)",
    col = c("lightblue", "lightpink"))
dev.off()

# 5. Relationship Analysis
cat("\n", paste(rep("=", 50), collapse=""), "\n")
cat("RELATIONSHIP ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

png(file.path(output_dir, "relationships_r.png"), width = 1500, height = 600, res = 300)
par(mfrow = c(1, 2))
boxplot(Age ~ Purchased, data = df, main = "Age Distribution by Purchase Status",
        xlab = "Purchased", ylab = "Age", names = c("No", "Yes"))
boxplot(EstimatedSalary ~ Purchased, data = df, 
        main = "Salary Distribution by Purchase Status",
        xlab = "Purchased", ylab = "Estimated Salary", names = c("No", "Yes"))
dev.off()

# Scatter plot
png(file.path(output_dir, "scatter_plot_r.png"), width = 1200, height = 800, res = 300)
plot(df$Age, df$EstimatedSalary, 
     col = ifelse(df$Purchased == 1, 
                  adjustcolor("red", alpha.f = 0.6), 
                  adjustcolor("blue", alpha.f = 0.6)),
     pch = 19,
     xlab = "Age", ylab = "Estimated Salary",
     main = "Age vs Estimated Salary (colored by Purchase Status)")
legend("topright", legend = c("No Purchase", "Purchase"), 
       col = c("blue", "red"), pch = 19)
dev.off()

# 6. Correlation Analysis
cat(paste(rep("=", 50), collapse=""), "\n")
cat("CORRELATION ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

numeric_df <- df[, c("Age", "EstimatedSalary", "Purchased")]
correlation_matrix <- cor(numeric_df)
cat("Correlation Matrix:\n")
print(correlation_matrix)
cat("\nCorrelation with Purchased:\n")
print(sort(correlation_matrix[, "Purchased"], decreasing = TRUE))

png(file.path(output_dir, "correlation_matrix_r.png"), width = 1000, height = 800, res = 300)
corrplot(correlation_matrix, method = "color", type = "upper",
         order = "hclust", tl.cex = 0.8, tl.col = "black",
         addCoef.col = "black", number.cex = 0.7)
dev.off()

cat("\n", paste(rep("=", 50), collapse=""), "\n")
cat("EDA COMPLETED SUCCESSFULLY!\n")
cat(paste(rep("=", 50), collapse=""), "\n")
